### PR TITLE
add a socks5 proxy

### PR DESCRIPTION
**Pull request by lynxcat**
Add a socks5 proxy to support cloud SSH login to the edge.

Documentation： [doc](https://github.com/kubeedge/edgemesh/issues/204)

Usage：
```
   ssh -o "ProxyCommand nc --proxy-type socks5 --proxy {$edgemesh-ip}:10800 {$nodeName} {$port}" root@any-ip

```

edgemesh IP defaults to 169.254.96.16

### CODE REVIEW

File: agent/pkg/proxy/socks5_proxy.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+	defer conn.Close()
+
+	var b [1024]byte
+	n, err := conn.Read(b[:])
+	if err != nil {
+		klog.Errorf("Unable to get data from client connection, err: %v", err)
+		return
+	}
+
+	if b[0] == 0x05 {
```
Comment (Poorunga) [inline agent/pkg/proxy/socks5_proxy.go:82]: Should we consider defining variables instead of hard coding here?

Comment (lynxcat) [reply agent/pkg/proxy/socks5_proxy.go]: Thank you for your suggestion. I have optimized the code.

File: agent/pkg/proxy/socks5_proxy.go [hunk 2]
```diff
@@ -268,13 +264,13 @@ func proxyConnectToRemote(host string, targetIP string, port int32, conn net.Con
 }
 
 func (s *Socks5Proxy) getTargetIpByNodeName(nodeName string) (targetIP string, err error) {
-	pods, err := s.kubeClient.CoreV1().Pods(constants.EdgeMeshNamespace).List(context.Background(), metav1.ListOptions{FieldSelector: "spec.nodeName=" + nodeName, LabelSelector: labelKubeedge})
+	pods, err := s.kubeClient.CoreV1().Pods(constants.EdgeMeshNamespace).List(context.Background(), metav1.ListOptions{FieldSelector: "spec.nodeName=" + nodeName, LabelSelector: LabelKubeedge})
```
Comment (Poorunga) [inline agent/pkg/proxy/socks5_proxy.go:267]: Use controller.APIConn.GetPodLister() instead

File: agent/pkg/proxy/config/config.go [hunk 3]
```diff
@@ -13,11 +20,22 @@ type EdgeProxyConfig struct {
...
 	// default 40001
 	ListenPort int `json:"listenPort,omitempty"`
+	//default 10800
+	SocksListenPort int `json:"socksListenPort,omitempty"`
+	//host nodename
+	NodeName string `json:"nodeName,omitempty"`
 }
 
 func NewEdgeProxyConfig() *EdgeProxyConfig {
+	nodeName, isExist := os.LookupEnv(meshConstants.MY_NODE_NAME)
```
Comment (Poorunga) [inline agent/pkg/proxy/config/config.go:30]: use common/util/util.go(*FetchNodeName) instead

**Comment by Poorunga** [comment]
@lynxcat  CI has a problem, let me restart your pr

/close

**Comment by khalid-jobs** [comment]
I think this feature shoud can be disable or enable. We should add a `enable` filed for this feature in edgemesh configuration.

File: agent/pkg/proxy/config/config.go [hunk 4]
```diff
@@ -13,11 +17,18 @@ type EdgeProxyConfig struct {
 	// ListenPort indicates the listen port of edgeproxy
 	// default 40001
 	ListenPort int `json:"listenPort,omitempty"`
+	//default 10800
```
Comment (khalid-jobs) [inline agent/pkg/proxy/config/config.go:20]:
```suggestion
	// default 10800
```

File: agent/pkg/proxy/config/config.go [hunk 5]
```diff
@@ -13,11 +17,18 @@ type EdgeProxyConfig struct {
...
+	SocksListenPort int `json:"socksListenPort,omitempty"`
+	//host nodename
```
Comment (khalid-jobs) [inline agent/pkg/proxy/config/config.go:22]:
```suggestion
	// host nodename
```

File: agent/pkg/proxy/module.go [hunk 6]
```diff
@@ -49,6 +50,12 @@ func newEdgeProxy(c *config.EdgeProxyConfig, ifm *informers.Manager) (proxy *Edg
 		return proxy, fmt.Errorf("new proxier err: %v", err)
 	}
 
+	// get socks5 proxy
+	proxy.Socks5Proxy, err = NewSocks5Proxy(listenIP, proxy.Config.SocksListenPort, proxy.Config.NodeName,  ifm.GetKubeClient())
+	if err != nil {
+		return proxy, fmt.Errorf("new socks5Proxy err: %v", err)
```
Comment (khalid-jobs) [inline agent/pkg/proxy/module.go:56]:
```suggestion
		return proxy, fmt.Errorf("new socks5Proxy err: %w", err)
```

File: agent/pkg/proxy/socks5_proxy.go [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+
+	_, err = conn.Write([]byte{Version, Success})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *SocksHandle) NewRequest(conn net.Conn) (err error) {
```
Comment (khalid-jobs) [inline agent/pkg/proxy/socks5_proxy.go:121]: If our method is copied from other github project and update for our own project, we should add a link 
```suggestion
# copy from https://github.com/txthinking/socks5/blob/e03c1217a50bd1363a2aaf58290da622256704fa/server_side.go#L125 and update
func (s *SocksHandle) NewRequest(conn net.Conn) (err error) {
```

File: agent/pkg/proxy/socks5_proxy.go [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+	ATYPIPv6 byte = 0x04 // 16 octets
+
+	// CmdConnect is connect command
+	CmdConnect byte = 0x01
+)
+
+// DefaultResponse is Socks5 returns data by default
+var DefaultResponse = []byte{Version, Success, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}
+
+type Request struct {
```
Comment (khalid-jobs) [inline agent/pkg/proxy/socks5_proxy.go:47]:
```suggestion
# copy form https://github.com/txthinking/socks5/blob/e03c1217a50bd1363a2aaf58290da622256704fa/socks5.go#from L86 and update
type Request struct {
```

Comment (lynxcat) [reply agent/pkg/proxy/socks5_proxy.go:48]: thanks, I've modified them.

**Comment by khalid-jobs** [comment]
Please make sure all  ci checks pass. And would you pull a request which includes the proposal doc (includes the purpose and architecure of this feature,[link](https://github.com/kubeedge/edgemesh/tree/main/docs/proposal)) and example doc (includes feature description and example ) [link](https://github.com/khalid-jobs/edgemesh/tree/main/docs/advanced) ?

**Comment by lynxcat** [comment]
> Please make sure all ci checks pass. And would you pull a request which includes the proposal doc (includes the purpose and architecure of this feature,[link](https://github.com/kubeedge/edgemesh/tree/main/docs/proposal)) and example doc (includes feature description and example ) [link](https://github.com/khalid-jobs/edgemesh/tree/main/docs/advanced) ?

Thank you. I fixed the lint error. I will write a new PR later to submit documents and modify the helm related configuration.

Here is the proposal document： [proposal](https://github.com/kubeedge/edgemesh/issues/204)

**Pull request merged**